experiments/2018-05-23-rust-on-android-JNA.md

```diff
@@ -0,0 +1,198 @@
+
+In is inside this interface that we will declare the headers that will link to our Rust library. We have only one external function to declare here, the `rust_greeting` function. Add the following interface declaration to your `JNA` interface.
+
+```
+String rust_greeting(String pattern);
+```
+
+That's it. We can now create some code that will call it.
+
+
+As with iOS, we're going to create a wrapper class to wrap the JNA binding and ensure that we have a more ergonmic Java API. In the project explorer on the left hand side of the studio window, ensure that `app > java > <domain>.greetings` is highlighted then go to `File > New > Java Class`. Name your class `RustGreetings` and click `OK`.
+
+![Create new class](assets/2017-09-21-rust-on-android/new-class.png)
+
+In your new class file, add the following code. Here we are defining the native interface to our Rust library and calling it `greeting`, with the same signature. The `sayHello` method simply makes a call to that native function.
+
+```java
+public class RustGreetings {
+
+    public String sayHello(String to) {
+        return JNA.INSTANCE.rust_greeting(to);
+    }
+}
+```
+
+We need to load our Rust library when the app starts, so add the following lines below the class declaration and before the `onCreate` method in `GreetingsActivity.java`.
+
+```java
+   static {
+        System.loadLibrary("greetings");
+    }
+```
+
+This looks for a library called `greetings.so` inside the jniLibs directory and picks the correct one for the current architecture.
+
+Open `res/layout/activity-greetings.xml`. In the `Component Tree` panel, highlight the `TextField` and open the `Properties` panel. Change the `ID` in the `Properties` panel to `greetingField`. This is how we are going to refer to it from our Activity.
+
+![Amend activity xml](assets/2017-09-21-rust-on-android/activity-greetings.png)
+
+Reopen `GreetingsActivity.java` and amend the `onCreate` method to call our greetings function and set the text on the `greetingField` `TextField` to the response value.
+
+```java
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_greetings);
+
+        RustGreetings g = new RustGreetings();
+        String r = g.sayHello("world");
+        ((TextView)findViewById(R.id.greetingField)).setText(r);
+    }
+```
+
+Build and run the app. If this is your first time in Android Studio, you may need to set up a simulator. When choosing/creating your simulator pick one with API 26. When the app starts, `Hello world` will be printed on your screen.
```

Comment:
It'd be helpful to explain why the emulator should use API 26 (and the app should require API 26). API 26 isn't available for ARM images, and I've been able to test on an ARM emulator successfully with API 25. The x86 emulator is much faster, of course, and API 26 is available for that. I'm just curious about the reasoning behind the choice of API level, and I suspect your other readers will be as well!